substrate/frame/delegated-staking/src/mock.rs

```diff
@@ -0,0 +1,319 @@
+			),
+		];
+
+		let _ = pallet_staking::GenesisConfig::<T> {
+			stakers: stakers.clone(),
+			// ideal validator count
+			validator_count: 2,
+			minimum_validator_count: 1,
+			invulnerables: vec![],
+			slash_reward_fraction: Perbill::from_percent(10),
+			min_nominator_bond: ExistentialDeposit::get(),
+			min_validator_bond: ExistentialDeposit::get(),
+			..Default::default()
+		}
+		.assimilate_storage(&mut storage);
+
+		let mut ext = sp_io::TestExternalities::from(storage);
+
+		ext.execute_with(|| {
+			// for events to be deposited.
+			frame_system::Pallet::<Runtime>::set_block_number(1);
+		});
+
+		ext
+	}
+	pub fn build_and_execute(self, test: impl FnOnce()) {
+		sp_tracing::try_init_simple();
+		let mut ext = self.build();
+		ext.execute_with(test);
+		ext.execute_with(|| {
+			DelegatedStaking::do_try_state().unwrap();
+		});
+	}
+}
+
+/// fund and return who.
+pub(crate) fn fund(who: &AccountId, amount: Balance) {
+	let _ = Balances::deposit_creating(who, amount);
+}
+
+/// Sets up delegation for passed delegators, returns total delegated amount.
+///
+/// `delegate_amount` is incremented by the amount `increment` starting with `base_delegate_amount`
+/// from lower index to higher index of delegators.
+pub(crate) fn setup_delegation_stake(
+	agent: AccountId,
+	reward_acc: AccountId,
+	delegators: Vec<AccountId>,
+	base_delegate_amount: Balance,
+	increment: Balance,
+) -> Balance {
+	fund(&agent, 100);
+	assert_ok!(DelegatedStaking::register_agent(RawOrigin::Signed(agent).into(), reward_acc));
+	let mut delegated_amount: Balance = 0;
+	for (index, delegator) in delegators.iter().enumerate() {
+		let amount_to_delegate = base_delegate_amount + increment * index as Balance;
+		delegated_amount += amount_to_delegate;
+
+		fund(delegator, amount_to_delegate + ExistentialDeposit::get());
+		assert_ok!(DelegatedStaking::delegate_to_agent(
+			RawOrigin::Signed(*delegator).into(),
+			agent,
+			amount_to_delegate
+		));
+	}
+
+	// sanity checks
+	assert_eq!(DelegatedStaking::stakeable_balance(&agent), delegated_amount);
+	assert_eq!(Agent::<T>::get(&agent).unwrap().available_to_bond(), 0);
+
+	delegated_amount
+}
+
+pub(crate) fn start_era(era: sp_staking::EraIndex) {
+	CurrentEra::<T>::set(Some(era));
+}
+
+pub(crate) fn eq_stake(who: AccountId, total: Balance, active: Balance) -> bool {
+	Staking::stake(&who).unwrap() == Stake { total, active } &&
+		get_agent(&who).ledger.stakeable_balance() == total
+}
+
+pub(crate) fn get_agent(agent: &AccountId) -> Agent<T> {
+	Agent::<T>::get(agent).expect("delegate should exist")
+}
+
+parameter_types! {
+	static ObservedEventsDelegatedStaking: usize = 0;
+}
+
+#[allow(unused)]
+pub(crate) fn events_since_last_call() -> Vec<crate::Event<Runtime>> {
+	let events = System::events()
+		.into_iter()
+		.map(|r| r.event)
+		.filter_map(
+			|e| if let RuntimeEvent::DelegatedStaking(inner) = e { Some(inner) } else { None },
+		)
+		.collect::<Vec<_>>();
```

Comment:
yes, let's use this :D

Reply:
And you can use this https://paritytech.github.io/polkadot-sdk/master/frame_system/pallet/struct.Pallet.html#method.read_events_for_pallet

Reply:
And note you have to set the block number to 1 to be able to read events.

Reply:
Yeah I set block number to 1 while building externalities.